Issue 20758: Changes to run telemetry/test_events.py on v6 topos

anders-nexthop opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/sonic-net/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
<!--
- Please include a summary of the change and which issue is fixed.
- Please also include relevant motivation and context. Where should reviewer start? background context?
- List any dependencies that are required for this change.
-->

Summary:

telemetry/test_events.py needs to be enhanced to run on ipv6 topos. `is_mgmt_ipv6_only` was added [here](https://github.com/sonic-net/sonic-mgmt/pull/20292), which can be used to check for ipv6-only topos. This can be used in this test to pass the correct ip address into the gnxi tool, which has already been enhanced to handle ipv6 addresses.

Changes were required to `is_bgp_state_idle()` in tests/common/devices/sonic.py to check for active neighbors in both address families.

The conditional skip has been removed.

Closes #20758 

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] New Test case
    - [ ] Skipped for non-supported platforms
- [x] Test case improvement


### Back port request
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [ ] 202405
- [ ] 202411
- [ ] 202505
- [ ] 202511

### Approach
#### What is the motivation for this PR?

Increase IPv6-only coverage.

#### How did you do it?

* extend bgp idle check to check ipv6 routes as well as ipv4
* use ipv6 mgmt addr in test_events.py when running on ipv6 topo
* remove conditional skip for test_events.py on ipv6 topo

#### How did you verify/test it?

Run the test on a virtual device with an ipv6-only topo. It fails before this change and passes after it.

#### Any platform specific information?

#### Supported testbed topology if it's a new test case?

### Documentation
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->

yanmo96 commented on tests/common/devices/sonic.py:
```diff
@@ -2200,23 +2200,26 @@ def is_bgp_state_idle(self):
-        return idle_count == (expected_idle_count - bgp_monitor_count)
+            if idle_count != (expected_idle_count - bgp_monitor_count):
+                return False
+        return True
```
Hi @bingwang-ms in ipv6-only testbed, we do not have ipv4 bgp neighbors

yxieca reviewed (dismissed):
AI agent on behalf of Ying. Reviewed; no issues found.

anders-nexthop committed:
Issue 20758: Changes to run telemetry/test_events.py on v6 topos

* use duthost.get_mgmt_ip() in generate_client_cli() to pick the right mgmt
  ip (v4 or v6) for the gnxi tool
* fix get_mgmt_ip() to read from the kernel via 'ip addr' rather than from
  SONiC's 'show ip[v6] interface', which filters out site-local v6 addresses
  (fec0::/10) and breaks the helper on v6-only mgmt setups; also skip
  link-local (fe80::/10) explicitly so the v6 regex never matches the
  '0/64' tail of 'eth0/64'
* remove the IPv6-only conditional skip and xfail entries for test_events

Signed-off-by: Anders Linn <anders@nexthop.ai>

mssonicbld commented:
/azp run

anders-nexthop commented:
Thanks for the review! All three items addressed in the rebase. The branch now contains a single commit on top of latest `master`.

**1. Dropped `is_bgp_state_idle()` change in `sonic.py`**

Master already has the equivalent v4+v6 fix. Removed entirely.

**2. `generate_client_cli()` now uses `duthost.get_mgmt_ip()`**

Replaced the `dut_basic_facts()` + `mgmt_ipv6` logic with the existing helper:

```python
mgmt_ip = duthost.get_mgmt_ip()["mgmt_ip"]
cmd = cmdFormat.format(mgmt_ip, env.gnmi_port, ...)
```

Same pattern that's already used in `tests/common/fixtures/tacacs.py`.

**3. Removed both `xfail` and `skip` entries**

The `xfail` block (added after this PR was opened) and the original `skip` block have both been removed from `tests_mark_conditions.yaml` for `telemetry/test_events.py::test_events`. Other "dut ipv6 mgmt ip not supported" entries are intentionally left alone in this PR; will follow up separately to verify and remove those as a batch.

**Verification (run on an IPv6 VTB with the 202505 SONiC image):**

- `test_events_cache_overflow[vlab-01]` — **PASSED**
- `test_events[vlab-01]::host_events` sub-test — **PASSED**, exercising `listen_for_events` → `generate_client_cli` with our `get_mgmt_ip()` change. The constructed gnmi command (`py_gnmicli.py -g -t <mgmt_ip> -p 50052 ...`) received responses from the DUT successfully, confirming the fix works end-to-end.
- `test_events[vlab-01]::dhcp-relay_events` sub-test — failed with `Unable to find vlan for test`. This is a pre-existing issue with the v6 skip introduced by #21649 and is not affected by this PR.

*— anders-bot (AI-assisted, on behalf of @anders-nexthop)*

anders-nexthop commented:
> Thanks for the review! All three items addressed in the rebase. The branch now contains a single commit on top of latest `master`.
> 
> **1. Dropped `is_bgp_state_idle()` change in `sonic.py`**
> 
> Master already has the equivalent v4+v6 fix. Removed entirely.
> 
> **2. `generate_client_cli()` now uses `duthost.get_mgmt_ip()`**
> 
> Replaced the `dut_basic_facts()` + `mgmt_ipv6` logic with the existing helper:
> 
> ```python
> mgmt_ip = duthost.get_mgmt_ip()["mgmt_ip"]
> cmd = cmdFormat.format(mgmt_ip, env.gnmi_port, ...)
> ```
> 
> Same pattern that's already used in `tests/common/fixtures/tacacs.py`.
> 
> **3. Removed both `xfail` and `skip` entries**
> 
> The `xfail` block (added after this PR was opened) and the original `skip` block have both been removed from `tests_mark_conditions.yaml` for `telemetry/test_events.py::test_events`. Other "dut ipv6 mgmt ip not supported" entries are intentionally left alone in this PR; will follow up separately to verify and remove those as a batch.
> 
> **Verification (run on an IPv6 VTB with the 202505 SONiC image):**
> 
> * `test_events_cache_overflow[vlab-01]` — **PASSED**
> * `test_events[vlab-01]::host_events` sub-test — **PASSED**, exercising `listen_for_events` → `generate_client_cli` with our `get_mgmt_ip()` change. The constructed gnmi command (`py_gnmicli.py -g -t <mgmt_ip> -p 50052 ...`) received responses from the DUT successfully, confirming the fix works end-to-end.
> * `test_events[vlab-01]::dhcp-relay_events` sub-test — failed with `Unable to find vlan for test`. This is a pre-existing issue with the v6 skip introduced by [Fix telemetry/test_events.py for v6 topos #21649](https://github.com/sonic-net/sonic-mgmt/pull/21649) and is not affected by this PR.
> 
> _— anders-bot (AI-assisted, on behalf of @anders-nexthop)_

@wangxinbot @yxieca please take a look at the updated PR, it needs to be re-approved after handling review comments.

mssonicbld commented:
/azp run

anders-nexthop commented:
Update — pushed one more change to this PR (force-pushed the same commit). While verifying on a true `is_mgmt_ipv6_only==True` setup, I found that `duthost.get_mgmt_ip()` (the helper @yxieca asked us to use) is itself broken on v6-only mgmt, so the previous version of this PR didn't actually fix the bug it claims to fix.

**What I did**

Modified `tests/common/devices/sonic.py::get_mgmt_ip()` to read the management interface IPs from the kernel via `ip -[4|6] -o addr show eth0` instead of SONiC's `show ip[v6] interface`. The v6 lookup explicitly skips `scope link` so we never look at `fe80::` addresses. This matches the source already used by `ansible/library/dut_basic_facts.py` to compute `is_mgmt_ipv6_only`, so the two stay consistent.

**Why it's needed**

Two pre-existing issues stack up on a v6-only mgmt DUT:

1. **`show ipv6 interface` filters out site-local addresses (`fec0::/10`).** That's where `is_mgmt_ipv6_only==True` test setups put the mgmt IP. The SONiC display layer returns only the link-local line:
   ```
   eth0    fe80::5054:ff:feb1:6b01%eth0/64    up/up    N/A    N/A
   ```
2. **The v6 regex `([a-fA-F0-9:]+)/\d+` is too loose.** Against the link-local line above, the proper v6 address part (`fe80::5054:ff:feb1:6b01`) doesn't satisfy the regex because the next char is `%`, not `/`. The regex then scans forward and matches `0/64` from the literal `eth0/64` at the end. `match.group(1)` returns `"0"`.

End result: `get_mgmt_ip()` confidently returns `{"mgmt_ip": "0", "version": "v6"}` on a v6-only mgmt DUT, and the gnxi command becomes `py_gnmicli.py -t 0` → `ConnectionRefusedError`.

The fix has two halves: switching from `show ipv6 interface` to `ip -6 addr show` (so we can actually see site-local addresses), and filtering out `scope link` (so we never return the link-local fe80:: as mgmt). Tightening the regex alone wouldn't have helped, because the SONiC display doesn't expose the address we need.

**How I found it**

After the rebase, I set up a true v6-only mgmt VTB:
- `t0-small-v6` topology with the 202505 SONiC image
- `sonic-mgmt-vm` container restarted with `--network=host` so it could reach `fec0::/64`
- Inventory `ansible_host` switched to the v6 address (`fec0::ffff:afa:1`)
- v4 removed from DUT eth0 (`sudo ip -4 addr del 10.250.0.101/24 dev eth0`)

In that environment `is_mgmt_ipv6_only` was correctly evaluated as `True`, but the `host_events` sub-test of `test_events` failed with the gnxi command using `-t 0` instead of the v6 address. Tracing through the helper turned up the two issues above.

**Verification (with the fix)**

On the same v6-only mgmt VTB:

```
py_gnmicli.py -g -t fec0::ffff:afa:1 -p 50052 -m subscribe ...
```

- `test_events_cache_overflow[vlab-01]` — **PASSED**
- `test_events[vlab-01]::host_events` — **PASSED**
- `test_events[vlab-01]::dhcp-relay_events` — still fails with `Unable to find vlan for test`, the pre-existing #21649 issue, unaffected by this PR

*— anders-bot (AI-assisted, on behalf of @anders-nexthop)*

yxieca commented:
Review note (non-blocking):
- get_mgmt_ip() now uses `ip -6 -o addr show eth0 | grep -v scope link`, which can return multiple v6 addresses. The regex picks the first match, which might be unexpected if both global/ULA exist. Consider filtering for scope global (or otherwise disambiguating).

Reviewed by AI agent on behalf of Ying.

anders-nexthop commented on tests/telemetry/telemetry_utils.py:
```diff
@@ -125,7 +125,12 @@ def generate_client_cli(duthost, gnxi_path, method=METHOD_GET, xpath="COUNTERS/E
+
+    dut_facts = duthost.dut_basic_facts()['ansible_facts']['dut_basic_facts']
+    is_mgmt_ipv6_only = dut_facts.get('is_mgmt_ipv6_only', False)
+    mgmt_ip = duthost.mgmt_ipv6 if is_mgmt_ipv6_only else duthost.mgmt_ip
```
We are now using `get_mgmt_ip()`

anders-nexthop commented:
> Review note (non-blocking):
> 
> * get_mgmt_ip() now uses `ip -6 -o addr show eth0 | grep -v scope link`, which can return multiple v6 addresses. The regex picks the first match, which might be unexpected if both global/ULA exist. Consider filtering for scope global (or otherwise disambiguating).
> 
> Reviewed by AI agent on behalf of Ying.

@yxieca @bingwang-ms I would like to address this issue in a follow-up commit. With these changes, the machinery is in place to remove the skips on several other telemetry tests. I'd rather get this one merged since it's been in the works for a long time, and then submit another which removes all of those skips (after I verify that they all still pass). Can we get this one approved and merged up? The skip-expiry just hit for this: https://github.com/sonic-net/sonic-mgmt/issues/20758#issuecomment-4358135883
